Fix Visitor Data Producing Invalid Tokens Due to URL Decoding

WillKirkmanM opened the pull request:
During the process of grabbing the authentication tokens from YouTube's API, there is an issue where the `Visitor Data` token is returned with URL encoded padding characters (**%3D**) which is the ASCII representation of the equals character (**=**).

The fix is decoding the URL through python's standard library; `urllib`, we use it's `.parse.unquote()` function and then strip the characters from the end using `rstrip('=')`.

WillKirkmanM committed:
Fix(extractor.py): URL Decoding for Visitor Data Token

unixfox commented:
How comes this is an issue? The visitordata used in the official www.youtube.com web interface, also ends with `%3D`

WillKirkmanM commented:
If you use the one ending in %3D, Invidious does not work.

unixfox commented:
Reviewing the PR, that's very odd. Could you do a video recording of installing invidious normally with docker: https://docs.invidious.io/installation/#docker then make the comparison with `%3D` added or removed.

This will help me a lot to understand what's going on. Thank you

WillKirkmanM commented:
Does your installation of invidious allow 1080p / 4K video playback with tokens that have %3D suffixed?

unixfox commented:
> Does your installation of invidious allow 1080p / 4K video playback with tokens that have %3D suffixed?

How this has to do with 1080p video playback?

WillKirkmanM commented:
Invidious does not allow you to play 1080p / 4k (anything above 360p) unless you have the "DASH (adaptive quality)" setting as it defaults to medium and requires those two tokens to enable the higher quality playback.

unixfox commented:
But you need these two tokens for medium too. This is why I don't get it.

You are trying to say that DASH does not work without the `%3D` suffix removed but medium still works with `%3D` suffix?

WillKirkmanM commented:
Yes, my main goal is to be able to watch videos at normal quality (1080p), initially, I was only able to watch them at 360p and then discovered that DASH allows it but only if you have the two tokens.

When generating the tokens with this tool, it did not work initially until i removed the `%3D` suffix. When I removed the `%3D` suffix, I was able to watch the videos with the normal (1080p) quality with DASH

unixfox commented:
Ok! Great. Thank you for the feedback, now I understand much better the issue.

unixfox commented:
Hello,
I'm going to close this issue since I just tested and dash works fine with a visitordata that ends with `%3D` generated by default by youtube-trusted-session-generator.

On www.youtube.com, the visitordata also ends with `%3D`. Since this program basically just extract the same value that YouTube themselves generated, I prefer to keep the behavior this way.

WillKirkmanM commented:
Perfect, I am on TrueNas and using invidious through docker and it did not work for me. Should it persist, I'll open an issue 👍 Merci pour lê voir.

dekeonus commented:
hmmm, I've just encountered this on my private instance:
po_token is **16** chars
visitor_data is *48* chars (including the %3D%3D without it is *42* chars)

```console
$ potoken-generator.py
visitor_data:  <42_chars_visitor_data>%3D%3D
po_token: <16_chars_po_token>
2025/04/28 00:38:23.861 [potoken] [WARNING] there is a high chance that the potoken generated won't work. Please try again on another internet connection
```

invidious ( 2025.03.14.0- @ master )
fails to play DASH with `visitor_data: <42_chars_visitor_data>%3D%3D`
succeeds to play DASH with `visitor_data: <42_chars_visitor_data>`

dekeonus commented:
I should also note that I used my system's site-packages:
*websockets-15.0.1*
*nodriver-0.45.2*
(noting that *undetected-chromedriver-3.5.5* was not installed at all, as it isn't a dependency of nodriver-0.39+ )

The browser used was google-chrome-stable
(This is a gentoo host - I really don't want to compile chromium, maybe if headless was supported I might build&install on the headless invidious host)
